plenum/server/monitor.py

```diff
@@ -31,42 +31,57 @@ class RequestMeasurement:
     Measure request params (throughput)
     """
 
-    def __init__(self, inner_window=15, min_cnt=16):
+    def __init__(self, throughput_window_size=15, throughput_min_cnt=16):
         self.reqs_in_window = 0
         self.throughput = 0
-        self.inner_window = inner_window
-        self.min_cnt = min_cnt
+        # map of client identifier and (total_reqs, avg_latency) tuple
+        self.avg_latencies = {}    # type: Dict(str, (int, float))
+        self.throughput_window_size = throughput_window_size
+        self.throughput_min_cnt = throughput_min_cnt
         self.first_ts = time.perf_counter()
         self.window_start_ts = self.first_ts
-        self.alpha = 2 / (self.min_cnt + 1)
+        self.alpha = 2 / (self.throughput_min_cnt + 1)
+
 
     def add_request(self, ordered_ts):
         self.update_time(ordered_ts)
         self.reqs_in_window += 1
 
-    def _moving_average(self, next_val):
+    def add_duration(self, identifier, duration):
+        if identifier not in self.avg_latencies:
+            self.avg_latencies[identifier] = (0, .0)
+        total_reqs, curr_avg_lat = self.avg_latencies[identifier]
+        total_reqs += 1
+        self.avg_latencies[identifier] = (total_reqs,
+                                          self._moving_average(curr_avg_lat,
+                                                               duration))
+
+    def _moving_average(self, p0, p1):
         """
         Implement exponential moving average
         """
-        return self.throughput * (1 - self.alpha) + next_val * self.alpha
+        return p0 * (1 - self.alpha) + p1 * self.alpha
 
     def update_time(self, current_ts):
-        while current_ts >= self.window_start_ts + self.inner_window:
-            self.throughput = self._moving_average(self.reqs_in_window / self.inner_window)
-            self.window_start_ts = self.window_start_ts + self.inner_window
+        while current_ts >= self.window_start_ts + self.throughput_window_size:
+            self.throughput = self._moving_average(self.throughput, self.reqs_in_window / self.throughput_window_size)
+            self.window_start_ts = self.window_start_ts + self.throughput_window_size
             self.reqs_in_window = 0
 
-    def get_avg_latency(self):
-        pass
+    def get_avg_latency(self, identifier):
+        if identifier not in self.avg_latencies:
+            return .0
+
+        return self.avg_latencies[identifier][1]
 
     def get_max_latency(self):
         pass
 
     def get_throughput(self, request_time):
-        if request_time < self.first_ts + (self.inner_window * self.min_cnt):
+        if request_time < self.first_ts + (self.throughput_window_size * self.throughput_min_cnt):
             return None
         self.update_time(request_time)
-        return self._moving_average(self.reqs_in_window / self.inner_window)
+        return self._moving_average(self.throughput, self.reqs_in_window / self.throughput_window_size)
```

Comment:
After some thinking I came to conclusion that if we want to get smooth averages between window boundaries it's not that simple, and what's done here is not helping in any way. Imagine you had throughput of 1, then you measure it before you get another request and this gives you values less than 1, which is yet constant until you get a request, after which throughput jumps back to 1 and remains here until next window starts, at which point it jumps below 1 again. So, I believe right thing to here is to either just return `self.throughput` or come up with better solutions (which can turn out quite complicated). I would just stick with simple options for now.